File: scanners/identity/auth_audit.py
"""
Authentication Audit Logger (Python) — Phase 14.2

Client-side and scanner authentication audit logger with tamper-evident SHA-256 hash chaining.
"""

from datetime import datetime, timezone
import hashlib
import json
from typing import Any, Dict, List, Optional
import uuid


class AuthAuditLogger:
    def __init__(self, initial_hash: str = "0" * 64):
        self.events: List[Dict[str, Any]] = []
        self.last_hash = initial_hash

    def sanitize_metadata(self, metadata: Dict[str, Any]) -> Dict[str, Any]:
        sanitized = {}
        for k, v in metadata.items():
            if any(term in k.lower() for term in ["password", "secret", "token", "auth", "key"]):
                sanitized[k] = "[REDACTED]"
            elif isinstance(v, dict):
                sanitized[k] = self.sanitize_metadata(v)
            else:
                sanitized[k] = v
        return sanitized

    def log_event(
        self,
        event_type: str,
        user_id: str = "anonymous",
        provider: str = "unknown",
        ip_address: str = "127.0.0.1",
        status: str = "SUCCESS",
        reason: str = "",
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        event_id = f"authevt_{uuid.uuid4()}"
        timestamp = datetime.now(timezone.utc).isoformat()
        clean_meta = self.sanitize_metadata(metadata or {})

        prev_hash = self.last_hash
        entry_data = json.dumps({
            "eventId": event_id,
            "eventType": event_type,
            "userId": str(user_id),
            "provider": str(provider),
            "ipAddress": str(ip_address),
            "status": str(status),
            "reason": str(reason),
            "metadata": clean_meta,
            "prevHash": prev_hash,
            "timestamp": timestamp,
        }, sort_keys=True)

        audit_hash = hashlib.sha256(entry_data.encode("utf-8")).hexdigest()
        self.last_hash = audit_hash

        record = {
            "eventId": event_id,
            "eventType": event_type,
            "userId": str(user_id),
            "provider": str(provider),
            "ipAddress": str(ip_address),
            "status": str(status),
            "reason": str(reason),
            "metadata": clean_meta,
            "prevHash": prev_hash,
            "auditHash": audit_hash,
            "timestamp": timestamp,
        }
        self.events.append(record)
        return record

    def verify_chain_integrity(self) -> Dict[str, Any]:
        expected_prev = "0" * 64
        for i, entry in enumerate(self.events):
            if i > 0 and entry["prevHash"] != expected_prev:
                return {"valid": False, "error": f"Chain broken at index {i}", "brokenIndex": i}

            entry_data = json.dumps({
                "eventId": entry["eventId"],
                "eventType": entry["eventType"],
                "userId": entry["userId"],
                "provider": entry["provider"],
                "ipAddress": entry["ipAddress"],
                "status": entry["status"],
                "reason": entry["reason"],
                "metadata": entry["metadata"],
                "prevHash": entry["prevHash"],
                "timestamp": entry["timestamp"],
            }, sort_keys=True)

            recalc = hashlib.sha256(entry_data.encode("utf-8")).hexdigest()
            if recalc != entry["auditHash"]:
                return {"valid": False, "error": f"Hash mismatch at index {i}", "brokenIndex": i}

            expected_prev = entry["auditHash"]

        return {"valid": True, "totalEvents": len(self.events)}

File: backend/src/identity/auth_audit.js
/**
 * Authentication Audit Logger — Phase 14.2
 *
 * Implements tamper-evident audit logging for enterprise identity events.
 * Uses SHA-256 cryptographic hash chaining for non-repudiation and integrity verification.
 * Enforces strict zero credential leakage (passwords, tokens, or client secrets are NEVER logged).
 */

const crypto = require("crypto");

const AUTH_EVENT_TYPES = {
  AUTH_SUCCESS: "AUTH_SUCCESS",
  AUTH_FAILURE: "AUTH_FAILURE",
  TOKEN_ISSUED: "TOKEN_ISSUED",
  TOKEN_REFRESHED: "TOKEN_REFRESHED",
  TOKEN_REVOKED: "TOKEN_REVOKED",
  SECRET_ROTATED: "SECRET_ROTATED",
  LOGOUT: "LOGOUT",
  SUSPICIOUS_REPLAY_DETECTED: "SUSPICIOUS_REPLAY_DETECTED",
  LDAP_BIND_ATTEMPT: "LDAP_BIND_ATTEMPT",
  AUTH_ACCOUNT_LOCKED: "AUTH_ACCOUNT_LOCKED",
  AUTH_MFA_CHALLENGE: "AUTH_MFA_CHALLENGE",
  AUTH_MFA_SUCCESS: "AUTH_MFA_SUCCESS",
  AUTH_MFA_FAILURE: "AUTH_MFA_FAILURE",
  PASSWORD_CHANGED: "PASSWORD_CHANGED",
};

class AuthAuditLogger {
  constructor(options = {}) {
    this.events = [];
    this.lastHash = options.initialHash || "0000000000000000000000000000000000000000000000000000000000000000";
    this.maxEvents = options.maxEvents || 10000;
  }

  /**
   * Sanitizes any object to guarantee zero password/token/secret leakage.
   */
  sanitizeMetadata(metadata = {}) {
    if (!metadata || typeof metadata !== "object") return {};
    const sanitized = {};
    for (const [key, value] of Object.entries(metadata)) {
      if (/(password|secret|token|authorization|apikey|code_verifier|privatekey)/i.test(key)) {
        sanitized[key] = "[REDACTED]";
      } else if (typeof value === "object" && value !== null) {
        sanitized[key] = this.sanitizeMetadata(value);
      } else {
        sanitized[key] = value;
      }
    }
    return sanitized;
  }

  /**
   * Logs an authentication event into the immutable hash chain.
   */
  logEvent({
    eventType,
    userId = "anonymous",
    provider = "unknown",
    ipAddress = "127.0.0.1",
    userAgent = "unknown",
    status = "SUCCESS",
    reason = "",
    metadata = {},
  }) {
    if (!AUTH_EVENT_TYPES[eventType]) {
      throw new Error(`Invalid authentication event type '${eventType}'`);
    }

    const eventId = `authevt_${crypto.randomUUID()}`;
    const timestamp = new Date().toISOString();
    const cleanMeta = this.sanitizeMetadata(metadata);

    const prevHash = this.lastHash;
    const entryData = JSON.stringify({
      eventId,
      eventType,
      userId: String(userId),
      provider: String(provider),
      ipAddress: String(ipAddress),
      userAgent: String(userAgent),
      status: String(status),
      reason: String(reason),
      metadata: cleanMeta,
      prevHash,
      timestamp,
    });

    const auditHash = crypto.createHash("sha256").update(entryData).digest("hex");
    this.lastHash = auditHash;

    const record = {
      eventId,
      eventType,
      userId: String(userId),
      provider: String(provider),
      ipAddress: String(ipAddress),
      userAgent: String(userAgent),
      status: String(status),
      reason: String(reason),
      metadata: cleanMeta,
      prevHash,
      auditHash,
      timestamp,
    };

    this.events.push(record);

    if (this.events.length > this.maxEvents) {
      this.events.shift();
    }

    return record;
  }

  /**
   * Verifies the cryptographic integrity of the audit chain.
   */
  verifyChainIntegrity() {
    let expectedPrev = "0000000000000000000000000000000000000000000000000000000000000000";

    for (let i = 0; i < this.events.length; i++) {
      const entry = this.events[i];
      if (i > 0 && entry.prevHash !== expectedPrev) {
        return {
          valid: false,
          error: `Hash chain broken at index ${i} (eventId: ${entry.eventId})`,
          brokenIndex: i,
        };
      }

      const entryData = JSON.stringify({
        eventId: entry.eventId,
        eventType: entry.eventType,
        userId: entry.userId,
        provider: entry.provider,
        ipAddress: entry.ipAddress,
        userAgent: entry.userAgent,
        status: entry.status,
        reason: entry.reason,
        metadata: entry.metadata,
        prevHash: entry.prevHash,
        timestamp: entry.timestamp,
      });

      const recalculated = crypto.createHash("sha256").update(entryData).digest("hex");
      if (recalculated !== entry.auditHash) {
        return {
          valid: false,
          error: `Tamper detected at index ${i} (eventId: ${entry.eventId}): hash mismatch`,
          brokenIndex: i,
        };
      }

      expectedPrev = entry.auditHash;
    }

    return { valid: true, totalEvents: this.events.length };
  }

  getRecentEvents(limit = 50) {
    return this.events.slice(-limit);
  }

  clear() {
    this.events = [];
    this.lastHash = "0000000000000000000000000000000000000000000000000000000000000000";
  }
}

// Global default singleton instance
const defaultAuthAuditLogger = new AuthAuditLogger();

module.exports = {
  AuthAuditLogger,
  AUTH_EVENT_TYPES,
  defaultAuthAuditLogger,
};
